### PR TITLE
Refactored Select internaly for Arrays to prepare for other changes

### PR DESCRIPTION
**Pull request by wuda-io**
I improved the internal structure of the component, because I saw that there where no Array used for the Options.
So there are no visible changes for the users, but the code is working now internally better and faster.
This is the preperation of dynamically loading Options without reinitializing the Component. But I also thougt
about making the Select similiar to the Autocomplete and give a "data" Object which can then be rendered.

**Future Plan:**
This should also be done in the Autocomplete. In the Autocomplete it will be a breaking change.
I want to make a combination of the Autocomplete and the Multi-Select with dynamic Loading Data
because I need it for my own Project, but I saw that it was already multiple times requested by other users.

For Reference see issues here:
#219 
#67
#115
#171
#124

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to change).

- [x] I have read the **[CONTRIBUTING document](https://github.com/materializecss/materialize/blob/main/CONTRIBUTING.md)**.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: js/select.js [hunk 1]
```diff
@@ -6,77 +6,37 @@
...
-     * @param {Object} options
-     */
     constructor(el, options) {
       super(FormSelect, el, options);
-
-      // Don't init if browser default version
-      if (this.$el.hasClass('browser-default')) {
-        return;
-      }
-
```
Comment (Smankusors) [inline js/select.js:27]: hmm why did you remove most of the comments there?

Comment (wuda-io) [reply js/select.js:27]: Because according to clean code, most of the comments were obsolete and had the same name as the functions, so i removed them to keep the code as clean as possible.

Comment (Smankusors) [reply js/select.js:27]: haha Clean Code, as I expected 😅 

but additional enters / blank lines are fine IMO, no need to remove them (ironically you also add additional blank lines 😄)

so far reading the code LGTM, but I haven't tried it on one of my apps.

Comment (Smankusors) [reply js/select.js:27]: wait.. uh oh... I just checked this and it somehow break my app 😅 

Though, it's kinda expected because I call the internal `_handleSelectChange` directly. Need some time to investigate this 🤕

Comment (wuda-io) [reply js/select.js:27]: Yes thats correct, I also saw that on my Project. The text is not updated correctly. I will investigate again this week an provide an additional commit which fixes this issue. Thank you for taking some time and test it. Looking forward for improvements. Best greets!

Comment (wuda-io) [reply js/select.js:27]: it should work now... i will test also in my project. You dont have to call the internal function but you can dispach a change Event: like `selectElem.dispatchEvent(new Event('change'))`

Comment (wuda-io) [reply js/select.js:27]: Yes! It works now in my Project too 🖖

Comment (wuda-io) [reply js/select.js:27]: I did another commit, because i now fixed an issue in the multi-select which was open.
I also improved the styling of the dropdown. It is now below the input. After this PR we can upgrade the Select with a Search function which would allow searching / ajax in the select and multi-select. The design change I made is also conform to the Material.io Specs from Google. The Dropdown is shown below the input. See this screenshot for details:
grafik

Comment (wuda-io) [reply js/select.js:27]: @Smankusors please check it in your project now. I tested it on my Project and it looks good now.
Looking forward for your response.

When should we do the next version iteration? Maybe 1.2.0 would be nice? What do you think @DanielRuf?
Then we could merge the other PRs too with breaking changes and then if it works for some time and things look good we could release a major 2.0.0.

Comment (DanielRuf) [reply js/select.js:27]: I think we can promote the current changes in the develop branch to `v1.1.0` since we only have an alpha release of it.

https://github.com/materializecss/materialize/commits/main?after=026fe0d666e7698a50192a34756fa4a3378d7cdf+69&branch=main

https://github.com/materializecss/materialize/tags

Can you coordinate this @Smankusors?
If you need some input from us, please let us know.

After this we can merge the current PRs and make a v1.2.0 release.

Comment (Smankusors) [reply js/select.js:27]: > @Smankusors please check it in your project now. I tested it on my Project and it looks good now. Looking forward for your response.

oh thanks, sorry if this takes days. Running out of time on weekend 😩 

I will test it now

> Can you coordinate this @Smankusors? If you need some input from us, please let us know.
> 
> After this we can merge the current PRs and make a v2.0.0-rc release.

wait, coordinate what? To release `v2.0.0-rc`?

File: js/select.js [hunk 2]
```diff
@@ -233,44 +237,41 @@
...
-      return isSelected;
+      const isSelected = this._isValueSelected(value);
+      if (isSelected) this._deselectValue(value);
+      else this._selectValue(value);
     }
-    _isOptionChosen(realOption) {
-      if (realOption.hasAttribute('disabled')) return false;
-      return realOption.selected || realOption.hasAttribute('selected');
+    _getSelectedOptions() {
+      return Array.prototype.map.call(this.el.selectedOptions, (realOption) => realOption);
```
Comment (Smankusors) [inline js/select.js]: eh wait, how does this work? 😅

File: js/select.js [hunk 3]
```diff
@@ -265,13 +265,21 @@
       else this._selectValue(value);
     }
     _getSelectedOptions() {
-      return Array.prototype.map.call(this.el.selectedOptions, (realOption) => realOption);
+      return Array.prototype.filter.call(this.el.selectedOptions, (realOption) => realOption);
```
Comment (Smankusors) [inline js/select.js:268]: eh wait, how does this work? 😅

File: js/select.js (hunk 2 above)
Comment (Smankusors) [reply js/select.js]: oops, sorry. Wrong file 😅

File: js/select.js (hunk 3 above)
Comment (wuda-io) [reply js/select.js:268]: https://stackoverflow.com/questions/11821261/how-to-get-all-selected-values-from-select-multiple-multiple

I first had troubles with setting things in the DOM. If you refresh your browser, the selected Values are kept in the cache and are not set in the DOM. So I fond the solution here. And since we drop support for IE it works good enough. If it is **filter** or **map** does not really matter it only converts the DataType to an Array.

Comment (wuda-io) [reply js/select.js:268]: https://caniuse.com/?search=selectedOptions

Comment (wuda-io) [reply js/select.js:268]: I think this could also be changed to `Array.from()` if you want.
https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/from
Have you tested it? When should we merge?

Comment (Smankusors) [reply js/select.js:268]: > If it is **filter** or **map** does not really matter it only converts the DataType to an Array.

I see. That's funny that selectedOptions is not an Array haha 😆

**Review by Smankusors — APPROVED** [review]
LGTM. Others? @materializecss/members

File: js/select.js (hunk 1 above)
Comment (DanielRuf) [reply js/select.js:27]: Afaik most of these commens are useful for autogenerating documentation from the annotations.

> wait, coordinate what? To release v2.0.0-rc?

No, we are still at v1.1-alpha. With coordinate I mean "feel free to merge if these changes are fine and make a v1.1 (without alpha) release".

Comment (DanielRuf) [reply js/select.js:27]: > I think we can promote the current changes in the develop branch to v1.1.0 since we only have an alpha release of it.

That was what I meant with "coordinate".
You can make a PR for the v1.1 release so the knowledge for doing releases is shared.

Comment (wuda-io) [reply js/select.js:27]: > Afaik most of these commens are useful for autogenerating documentation from the annotations.
> 
> > wait, coordinate what? To release v2.0.0-rc?
> 
> No, we are still at v1.1-alpha. With coordinate I mean "feel free to merge if these changes are fine and make a v1.1 (without alpha) release".

Yes but the names and descriptions where the same as the function names, so no advantage here. Also the documentation is handwritten for now so i thought i clean up the code and make it more compact. This can be merged from my point of view. Looking forward for some improvement of this project 👍

File: js/select.js (hunk 3 above)
Comment (wuda-io) [reply js/select.js:268]: Yes it is some DOM specific array which has to be converted for some odd reason...the same issue is with querySelectorAll Method.

**Comment by Smankusors** [comment]
I think if nobody objects about this, then lets merge this one

thanks btw @wuda-io 👍
